### DISCUSSION TITLE
API automation

### DISCUSSION THREAD

**Discussion by rakesh-308** [apache/apisix#7936] [General]
I noticed, in routes we configured there is a possibility to export them in OpenAPI specifications and use a UI tool like swagger to import them and use it as developer portal.

Is it possible to expose the documentation regarding endpoints of services/upstream?

Is there any way to automate the exposure documentation for new endpoints we expose in service (and that service is exposed by a route in Gateway)

In Kong, there is something called 'Developer Portal', I was trying to find something similar which can help exposing developers our apis documentation in a automated fashion, but could not see anything yet :(

**Comment by moonming** [comment]
> Is it possible to expose the documentation regarding endpoints of services/upstream?

can you give an example? like postman for API spec?

**Reply by rakesh-308** [reply]
Just like Kong has a developer portal, 

A way to automate documentation to world developers, all endpoints in a service (assuming service is exposed via a route)

**Reply by rakesh-308** [reply]
Kong has this https://docs.konghq.com/enterprise/2.4.x/developer-portal/

**Comment by daolfer** [comment]
Also interested in this feature. Do we have progress in this topic?

**Comment by ZiriusDeveloper** [comment]
It would be nice, if this feature included.  AWS API gateway has similar feature
We document all the paths, auth and integrations in yaml/json (openapi  standard) and import it in dashboard. So it creates necessary routes and maps plugins and authentications 
Same way we can export it to yaml / json in openapi standard. This makes life more easier :-)

**Reply by juzhiyuan** [reply]
Take a look at https://github.com/api7/apisix-cicd-demo (may help)

**Comment by sebgott** [comment]
I think being able to reference an OpenAPI spec in Kubernetes custom resources would be a nice feature to have instead of using the adc cli manually or defining every route in the spec. A lot of potential users are used to creating their gateways based on this for example in Azure API Management or Traefik. Something like this:

```yaml
apiVersion: apisix.apache.org/v2alpha1
kind: ApisixRoute
metadata:
  name: foo-bar-route
spec:
  configMapRef:
     name: foo-bar-api-spec
     key: openapi.yaml
   upstreams:
    - name: foo-bar-upstream
```
